---
  title: "Genome Annotation Project"
subtitle: "Figure 4: genic vs. intergenic random forest Gini index"
date: "April 10, 2024"
author: "Adam Freedman"
output: 
  html_document: default
---
  
## Set up
```{r,,message=FALSE, warning=FALSE}
currentdir=getwd()
setwd(currentdir)
library(tidyverse)
``` 

## Data
```{r}
importance_data <- read_table("genic_intergenic_importance_concat.tsv")
importance_data <- importance_data %>% mutate(predictor=factor(predictor,levels=c("blastphit","GCcontent","totalrepprop","length","maxTPM",
                              "medianTPM","fractTPMlessThan1","number_exons","Complete","StopCodon","StartCodon","NumInternalStops")))

importance_data <- importance_data %>% mutate(method=factor(method,levels=c("brakerprot","brakerrna","cgpprot","cgprna",
                                           "scalhisat","scalstar","stiehisat","stiestar","maker","toga")))

importance_data <- importance_data %>% mutate(species = str_replace(species, "dmel", "D. melanogaster")) %>%
    mutate(species = str_replace(species, "athaliana", "A. thaliana")) %>%
    mutate(species = str_replace(species, "zmays", "Z. mays")) %>% 
    mutate(species = str_replace(species, "chicken", "G. gallus")) %>%
    mutate(species = str_replace(species, "human", "H. sapiens"))

importance_data$species<- factor(importance_data$species,levels=c("D. melanogaster","G. gallus","H. sapiens","A. thaliana","Z. mays"))

methodlabels<-c(expression(BRAKER[protein]),expression(BRAKER[RNA]),expression(CGP[protein]),expression(CGP[RNA]),expression(Scallop[HISAT2]),expression(Scallop[STAR]),expression(Stringtie[HISAT2]),expression(Stringtie[STAR]),"MAKER","TOGA")

predictor_labels <- c("BLASTP match","GC content","Repeat fraction","Length","Maximum TPM",
                              "Median TPM","% samples TPM <1","# exons","Complete ORF","Stop codon","Start codon","# internal stops")

```

## Figure 4A: *H. sapiens*
```{r}
gini_dotplot_human <- importance_data %>% filter(species=="H. sapiens") %>% ggplot() + 
    theme_classic() + 
    geom_point(aes(x=method,y=MeanDecreaseGini,color=predictor),position = position_jitter(w = 0.18,h = 0),alpha=0.65,size=1.6) +
    theme(axis.text.x=element_text(angle=90, hjust=1,vjust=0.3,size=9)) +
    scale_color_manual(labels=predictor_labels,values=c("yellow2","#e69f00", "firebrick1","firebrick4", "lightskyblue", "blue","deepskyblue4","gold4","darkorchid3", "magenta","gray70","black")) +

    theme(legend.text.align = 0,legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) +
    ylab("Mean decrease in Gini index") + xlab("") +
    scale_x_discrete(breaks=levels(importance_data$method),labels=methodlabels) +
    guides(fill = guide_legend(byrow = TRUE)) +
    labs(color = "Predictor") + 
    theme(strip.text = element_text(face="italic"))
print(gini_dotplot_human)

ggsave(gini_dotplot_human,file="Figure_4A.pdf",width=4.75,height=4)
```

## Figure 4B: *Z. mays* 
```{r}
gini_dotplot_zmays <- importance_data %>% filter(species=="Z. mays") %>% ggplot() + 
    theme_classic() + 
    geom_point(aes(x=method,y=MeanDecreaseGini,color=predictor),position = position_jitter(w = 0.18,h = 0),alpha=0.65,size=1.6) +
    theme(axis.text.x=element_text(angle=90, hjust=1,vjust=0.3,size=9)) +
    scale_color_manual(labels=predictor_labels,values=c("yellow2","#e69f00", "firebrick1","firebrick4", "lightskyblue", "blue","deepskyblue4","gold4","darkorchid3", "magenta","gray70","black")) +

    theme(legend.text.align = 0,legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) +
    ylab("Mean decrease in Gini index") + xlab("") +
    scale_x_discrete(breaks=levels(importance_data$method),labels=methodlabels) +
    guides(fill = guide_legend(byrow = TRUE)) +
    labs(color = "Predictor") + 
    theme(strip.text = element_text(face="italic"))
print(gini_dotplot_zmays)

ggsave(gini_dotplot_zmays,file="Figure_4B.pdf",width=4.75,height=4)
```

```