---
  title: "Genome Annotation Project"
subtitle: "Figure 7: expression"
date: "April 10, 2024"
author: "Adam Freedman"
output: 
  html_document: default
---
  
## Set up
```{r,,message=FALSE, warning=FALSE}
currentdir=getwd()
setwd(currentdir)
library(tidyverse)
``` 

## Data
```{r,,message=FALSE, warning=FALSE}
annotations <-read_csv("GenomeAnnotationMetrics_final.csv")
annotations <- annotations %>% filter(group!="heliconines")
annotations <- annotations %>% filter(!method %in% c("BrkProt-unfilt","BrkProt-fullsupport","BrkRNA-unfilt",
                                                     "BrkRNA-fullsupport","CGPdenovo","TSEBRA","MakerNoRnaNoEvm",
                                                     "MakerRnaEvm","MakerNoRnaEvm")) 

genomeQC <-read_csv("GenomeQcMetrics_final.csv") 
data <- annotations<-left_join(annotations,genomeQC,by = join_by(species==Species))

data$method <- factor(data$method,levels=c("BrkProt","BrkRNA","CGPprotein","CGPrnaseq",
                                           "ScalHisat","ScalStar","StrHisat","StrStar","MakerRnaNoEvm","TOGA"))

data <- data %>% mutate(species = str_replace(species, "dmelanogaster", "D. melanogaster")) %>%
     mutate(species = str_replace(species, "dpseudoobscura", "D. pseudoobscura")) %>%  
     mutate(species = str_replace(species, "dyakuba", "D. yakuba")) %>%
     mutate(species = str_replace(species, "alyrata", "A. lyrata")) %>%
     mutate(species = str_replace(species, "athaliana", "A. thaliana")) %>%
     mutate(species = str_replace(species, "boleracea", "B. oleracea")) %>%
     mutate(species = str_replace(species, "crubella", "C. rubella")) %>%
     mutate(species = str_replace(species, "osativa", "O. sativa")) %>%
     mutate(species = str_replace(species, "sitalica", "S. italica")) %>%
     mutate(species = str_replace(species, "bdistachyon", "B. distachyon")) %>%
     mutate(species = str_replace(species, "zmays", "Z. mays")) %>%
     mutate(species = str_replace(species, "human", "H. sapiens")) %>%
     mutate(species = str_replace(species, "macaque", "M. mulatta")) %>%
     mutate(species = str_replace(species, "dog", "C. familiaris")) %>%
     mutate(species = str_replace(species, "mouse", "M. musculus")) %>%
     mutate(species = str_replace(species, "chicken", "G. gallus")) %>%
     mutate(species = str_replace(species, "quail", "C. japonica")) %>%
     mutate(species = str_replace(species, "mallard", "A. platyrhynchos"))

data <- data %>% mutate_at(vars(species),factor,levels=c("D. melanogaster","D. pseudoobscura","D. yakuba",
                                                         "A. thaliana", "A. lyrata","B. oleracea","C. rubella",
                                                         "Z. mays","O. sativa","S. italica","B. distachyon",
                                                         "G. gallus","A. platyrhynchos","C. japonica","H. sapiens",
                                                         "M. mulatta","C. familiaris","M. musculus"))

methodlabels<-c(expression(BRAKER[protein]),expression(BRAKER[RNA]),expression(CGP[protein]),
                expression(CGP[RNA]),expression(Scallop[HISAT2]),expression(Scallop[STAR]),
                expression(Stringtie[HISAT2]),expression(Stringtie[STAR]),"MAKER","TOGA")
```

### Figure 7A: low TPM gene frequencies
```{r}
methodlabels<-c(expression(BRAKER[protein]),expression(BRAKER[RNA]),expression(CGP[protein]),expression(CGP[RNA]),expression(Scallop[HISAT2]),expression(Scallop[STAR]),expression(Stringtie[HISAT2]),expression(Stringtie[STAR]),"MAKER","TOGA")

gene_lowtpm_plot <- data %>% ggplot(aes(x=species,y=gene_TPMlthan1_noUTR)) + geom_point(aes(fill=method),color="black",pch=21,position = position_jitter(w = 0.18,h = 0),alpha=0.7,size=2) +
    scale_fill_manual(labels=methodlabels,values=c("yellow2","#e69f00", "firebrick1","firebrick4", "lightskyblue", "royalblue3","darkorchid3", "magenta","gray76","#000000")) + 
theme(legend.text.align = 0) + 
theme(axis.text.x=element_text(angle=90, hjust=1,vjust=0.3,face="italic")) + 
ylab("Proportion genes median TPM < 1") +
xlab("") + 
labs(fill = "Method") +
geom_vline(xintercept=3.5,linetype="solid") +
geom_vline(xintercept=7.5,linetype="solid")+
geom_vline(xintercept=11.5,linetype="solid") +
geom_vline(xintercept=14.5,linetype="solid") 

print(gene_lowtpm_plot)
ggsave(gene_lowtpm_plot,file="global_gene_lowtpm_plot.pdf",width=6,height=4)
```

### Figure 7B: alignment rates
```{r}
methodlabels<-c(expression(BRAKER[protein]),expression(BRAKER[RNA]),expression(CGP[protein]),expression(CGP[RNA]),expression(Scallop[HISAT2]),expression(Scallop[STAR]),expression(Stringtie[HISAT2]),expression(Stringtie[STAR]),"MAKER","TOGA")

alignrate_noutr_plot <- data %>% filter(group!="heliconines") %>% ggplot() +
geom_point(aes(x=species,y=RefAlignRateNoUTR),shape=4,color="black",size=3.5) +
geom_point(aes(x=species,y=median_annot_align_noutr,fill=method),color="black",pch=21,position = position_jitter(w = 0.2,h = 0),alpha=0.7,size=2) +
scale_fill_manual(labels=methodlabels,values=c("yellow2","#e69f00", "firebrick1","firebrick4", "lightskyblue", "royalblue3","darkorchid3", "magenta","gray76","#000000")) + 
theme(legend.text.align = 0) + 
theme(axis.text.x=element_text(angle=90, hjust=1,vjust=0.3,face="italic")) + 
ylab(expression(paste(RNA[seq]," alignment rate"))) +
xlab("") + 
guides(fill = guide_legend(title = "Method")) +
geom_vline(xintercept=3.5,linetype="solid") +
geom_vline(xintercept=7.5,linetype="solid")+
geom_vline(xintercept=11.5,linetype="solid") +
geom_vline(xintercept=14.5,linetype="solid") 


print(alignrate_noutr_plot)
ggsave(alignrate_noutr_plot,file="alignrate_noutr_plot.pdf",width=6,height=4)
```