---
  title: "Genome Annotation Project"
subtitle: "Figure 1: proteome compleasm barplots: 
date: "December 13, 2024"
author: "Adam Freedman"
output: 
  html_document: default
---
  
## Set up
```{r,,message=FALSE, warning=FALSE}
currentdir=getwd()
setwd(currentdir)
library(tidyverse)
``` 

## Data
```{r,,message=FALSE, warning=FALSE}
annotations <-read_csv("../../data/Table_S3.csv")
annotations <- annotations %>% filter(!method %in% c("BrkProt-unfilt","BrkProt-fullsupport","BrkRNA-unfilt",
                                                     "BrkRNA-fullsupport","CGPdenovo","TSEBRA","MakerNoRnaNoEvm",
                                                     "MakerRnaEvm","MakerNoRnaEvm")) 

methodlabels<-c("BRAKER2 (P)","BRAKER1 (P)","BRAKER3 (P)",expression(paste(CGP[protein]," (P)")),expression(paste(CGP[RNA]," (P)")),expression(Scallop[HISAT2]),expression(Scallop[STAR]),expression(Stringtie[HISAT2]),expression(Stringtie[STAR]),"MAKER (P)","TOGA (P)","Liftoff")

annotations$method <- factor(annotations$method,levels=c("BrkProt","BrkRNA","Braker3","CGPprotein","CGPrnaseq",
                                           "ScalHisat","ScalStar","StrHisat","StrStar","MakerRnaNoEvm","TOGA","liftoff"))

annotations <- annotations %>% mutate(species = str_replace(species, "dmelanogaster", "D. melanogaster")) %>%
     mutate(species = str_replace(species, "dpseudoobscura", "D. pseudoobscura")) %>%  
     mutate(species = str_replace(species, "dyakuba", "D. yakuba")) %>%
     mutate(species = str_replace(species, "alyrata", "A. lyrata")) %>%
     mutate(species = str_replace(species, "athaliana", "A. thaliana")) %>%
     mutate(species = str_replace(species, "boleracea", "B. oleracea")) %>%
     mutate(species = str_replace(species, "crubella", "C. rubella")) %>%
     mutate(species = str_replace(species, "osativa", "O. sativa")) %>%
     mutate(species = str_replace(species, "sitalica", "S. italica")) %>%
     mutate(species = str_replace(species, "bdistachyon", "B. distachyon")) %>%
     mutate(species = str_replace(species, "zmays", "Z. mays")) %>%
     mutate(species = str_replace(species, "bmori", "B. mori")) %>%
     mutate(species = str_replace(species, "herato", "H. erato")) %>%
     mutate(species = str_replace(species, "dplexippus", "D. plexippus")) %>%
     mutate(species = str_replace(species, "human", "H. sapiens")) %>%
     mutate(species = str_replace(species, "macaque", "M. mulatta")) %>%
     mutate(species = str_replace(species, "dog", "C. familiaris")) %>%
     mutate(species = str_replace(species, "mouse", "M. musculus")) %>%
     mutate(species = str_replace(species, "chicken", "G. gallus")) %>%
     mutate(species = str_replace(species, "quail", "C. japonica")) %>%
     mutate(species = str_replace(species, "mallard", "A. platyrhynchos"))


annotations <- annotations %>% mutate_at(vars(species),factor,levels=c("D. melanogaster",
              "D. yakuba", "D. pseudoobscura","A. thaliana", "A. lyrata","C. rubella","B. oleracea","Z. mays","S. italica","O. sativa",
               "B. distachyon", "G. gallus","C. japonica","A. platyrhynchos","H. sapiens","M. mulatta","M. musculus","C. familiaris","H. erato","D. plexippus","B. mori"))
```


## 1A: vertebrates

```{r}
vertebrates_compleasm_protein_barplot <- annotations %>% filter(group =="mammals" | group=="birds") %>% ggplot() + theme(text = element_text(size = 12))  +
    geom_bar(aes(species,y=compleasm_protein,fill=method),width=0.9,stat="identity",position="dodge") +
     scale_fill_manual(labels=methodlabels,values=c("yellow3","orange","brown4", "red","coral3", "lightskyblue", "royalblue3","darkorchid3", "magenta","gray72","green3","seagreen")) +
    theme(axis.text.x=element_text(angle=45, hjust=1,vjust=1,size=8,face="italic")) + 
    theme(axis.text.y=element_text(size=8)) + 
    theme(axis.title.y=element_text(size=10)) +
    ylab("Compleasm score") + xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
    theme(legend.text.align = 0,legend.text=element_text(size=12),legend.spacing.y = unit(0.1, "cm"),legend.key.size= unit(0.45, "cm")) + 
    guides(fill = guide_legend(byrow = TRUE)) +
    labs(fill = "") 
    


print(vertebrates_compleasm_protein_barplot)
ggsave(file="Figure_1A_2024.12.13.pdf",width=7,height=3)
```


## 1B: plants

```{r}
plants_proteome_compleasm_barplot <- annotations %>% filter(group =="rosids" | group=="monocots") %>% 
  mutate(compleasm_protein=replace_na(compleasm_protein,-1)) %>% 
  ggplot() + theme(text = element_text(size = 12))  +
    geom_bar(aes(species,y=compleasm_protein,fill=method),width=0.9,stat="identity",position="dodge") +
     scale_fill_manual(labels=methodlabels,values=c("yellow3","orange","brown4", "red","coral3", "lightskyblue", "royalblue3","darkorchid3", "magenta","gray72","green3","seagreen")) +
    theme(axis.text.x=element_text(angle=45, hjust=1,vjust=1,size=10,face="italic")) + 
    ylab("Compleasm score") + xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
    theme(legend.text.align = 0,legend.text=element_text(size=12),legend.spacing.y = unit(0.1, "cm"),legend.key.size= unit(0.45, "cm")) + 
    guides(fill = guide_legend(byrow = TRUE)) +
    labs(fill = "") 
    


print(plants_proteome_compleasm_barplot)
ggsave(file="Figure_1B_2024.12.13.pdf",width=9,height=3.42)
```


## 1C: dipterans

```{r}
insects_proteome_compleasm_barplot <- annotations %>% filter(group =="dipterans") %>% 
  mutate(compleasm_protein=replace_na(compleasm_protein,-1)) %>% ggplot() + theme(text = element_text(size = 10))  +
    geom_bar(aes(species,y=compleasm_protein,fill=method),width=0.9,stat="identity",position="dodge") +
    scale_fill_manual(labels=methodlabels,values=c("yellow3","orange","brown4", "red","coral3", "lightskyblue", "royalblue3","darkorchid3", "magenta","gray72","green3","seagreen")) +
    theme(axis.text.x=element_text(angle=45, hjust=1,vjust=1,size=8,face="italic")) + 
    ylab("Compleasm score") + xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
    theme(legend.text.align = 0,legend.text=element_text(size=12),legend.spacing.y = unit(0.1, "cm"),legend.key.size= unit(0.35, "cm")) + 
    guides(fill = guide_legend(byrow = TRUE)) +
    labs(fill = "") 
    


print(insects_proteome_compleasm_barplot)
ggsave(file="Figure_1C_2024.12.13.pdf",width=5,height=3.05)
```
