# Transcript assembly from RNA-seq read
In [Freedman and Sackton, 2024](https://www.biorxiv.org/content/10.1101/2024.04.12.589245v2), we evaluated genome annotation by assembling transcripts directly from RNA-seq reads. This involves two steps: spliced alignment of RNA-seq reads to genomes, and using those alignments to build graph structures from which transcripts assemblies will be inferred. We evaluated combinations of two aligners [Hisat2](http://daehwankimlab.github.io/hisat2/) and [STAR](https://github.com/alexdobin/STAR), with two transcript assemblers, [Stringtie2](https://ccb.jhu.edu/software/stringtie/) and [Scallop](https://github.com/Kingsford-Group/scallop) to perform sample-level assemblies, leading to four aligner-assembler combinations. 

For alignment with STAR, a 2-pass alignment strategy is implemented, in which splice sites detected across all samples in the 1st pass are used to inform alignment of individual samples in the 2nd pass. Sequence alignments from either aligner that are in *sam* format are then coordinated sorted and converted to *bam* format. Next, these bam files are passed to the RNA-seq transcript assemblers. Sample-level *Stringtie2* assemblies are then merged using the stringtie *--merge* argument, while sample-level assemblies generated with *Scallop* are merged using [TACO](http://tacorna.github.io/). A few important notes regarding the current workflow:
* We performed alignment and assembly in a strand-agnostic fashion, such that no strand-specific arguments have been employed. We did this to avoid spending an inordinate amount of time trying to confirm strand-confiuration for SRA accessions used in this study. 
* We used default settings for minimum amd maximum intron size and --sjdbOverhang, the latter of which is typically set to median sequence read length -1 (the default is 100bp, and the STAR developers believe it generally leads to good alignment performance).

## Snakemake
The generation of transcript assemblies from all four aligner-assembler combinations were generated using a snakemake workflow, found [here]().

