---
  title: "Genome Annotation Project"
subtitle: "Figure S2: CDS count ratio vs. length ratio for taxonomic groups: 
date: "December 13, 2024"
author: "Adam Freedman"
output: 
  html_document: default
---
  
## Set up
```{r,,message=FALSE, warning=FALSE}
currentdir=getwd()
setwd(currentdir)
library(tidyverse)
``` 

```{r}
classdata <-read_table("../../data/refspecies_methodclass_unique_intergenic_stats.tsv")
cor.test(classdata$classunique,classdata$classintergenic)
```
## Figure S4
```{r}
classdataplot<-classdata %>% 
  mutate(species = str_replace(species, "dmel", "D. melanogaster")) %>%
  mutate(species = str_replace(species, "zmays", "Z. mays")) %>%
  mutate(species = str_replace(species, "human", "H. sapiens")) %>%
  mutate(species = str_replace(species, "athaliana", "A. thaliana")) %>%
  mutate(species = str_replace(species, "mouse", "M. musculus")) %>%
  
  ggplot(aes(x=100*classunique,y=100*classintergenic,color=methodclass,shape=species)) +
  geom_point(size=3) +
  scale_color_manual(values=c("lift"="seagreen",braker="orange",cgp="firebrick",stie="darkorchid3",scal="royalblue",maker="gray72"),labels=c("lift"="liftover","stie"="Stringtie","braker"="BRAKER","cgp"="CGP","maker"="MAKER","scal"="Scallop")) +
ylab("Percent genes intergenic (relative to NCBI)") +
xlab("Percent genes unique to method class") +
guides(color = guide_legend(title = "Method class",override.aes = list(shape=15),order=1,label.theme = element_text(angle = 0, size=10,face = "plain"))) +
guides(shape = guide_legend(title = "Species",order=2,label.theme = element_text(angle = 0, size=10,face = "italic"))) +
guides(fill = guide_legend(title = "")) +
annotate("text",x=60,y=5,label=expression(paste("Pearson's ",rho,"=0.975, p=","2.2x","10"^"-16")),size=3)
  
classdataplot
ggsave(classdataplot,file="Figure_S4.pdf",width=5,height=5)
```