---
  title: "Genome Annotation Project"
subtitle: "Figures S16 - S19: method integration"
date: "May 28, 2024"
author: "Adam Freedman"
output: 
  html_document: default
---

## Set up
```{r,,message=FALSE, warning=FALSE}
setwd("/Users/adamfreedman/Dropbox/informatics/genome_annotation_evaluation/2023_plotsandcode/integration")
library(tidyverse)
library(scales)
```

```{r}
data <- read_csv("methodsintegration.csv")
```

## Figure S16: tringtie as base annotation


### Figure S16A: BUSCO scores
```{r}
methodlabels<-c(expression(Stringtie[STAR]),"TOGA",expression(BRAKER[protein]),"Stringtie+TOGA",expression(paste("Stringtie+TOGA+",BRAKER[protein])))
stie1st_busco <- data %>% filter(method %in% c("stiestar","toga","bprot","stietoga","stietogabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","toga","bprot","stietoga","stietogabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=busco_score,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","black","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("BUSCO score") + xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(stie1st_busco)
ggsave(stie1st_busco,file="Figure_S16A.pdf",width=4,height=2.5)
```

### Figures S16B: number genes with BLASTP hits
```{r}
stie1st_geneblast <- data %>% filter(method %in% c("stiestar","toga","bprot","stietoga","stietogabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","toga","bprot","stietoga","stietogabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=number_genes_blasthits,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","black","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("Genes with BLASTP hits to NCBI") + xlab("") + scale_y_continuous(expand=c(0,0)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(stie1st_geneblast)
ggsave(stie1st_geneblast,file="Figure_S16B.pdf",width=4,height=3)
```

### Figure S16C: fraction genes with NCBI BLASTP hits
```{r}
stie1st_blastfrac <- data %>% filter(method %in% c("stiestar","toga","bprot","stietoga","stietogabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","toga","bprot","stietoga","stietogabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=100*blastp_fraction,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","black","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("% of genes with NCBI BLASTP hits") +
      xlab("") + scale_y_continuous(expand=c(0,0)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(stie1st_blastfrac)
ggsave(stie1st_blastfrac,file="integration_blastphitfraction_stringtie1st.pdf",width=4,height=3)
```


### Figure S16D: RNA-seq alignment rate 
```{r}
stie1st_alignrate <- data %>% filter(method %in% c("stiestar","toga","bprot","stietoga","stietogabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","toga","bprot","stietoga","stietogabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=align_rate,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","black","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab(expression(paste(RNA[seq], " alignment rate"))) +
      xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,0.8)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(stie1st_alignrate)
ggsave(stie1st_alignrate,file="Figure_S16D.pdf",width=4,height=2.5)
```


## Figure S17: TOGA as base annotation

### Figure S17A: BUSCO scores

```{r}
methodlabels<-c(expression("TOGA", Stringtie[STAR]),expression(BRAKER[protein]),"TOGA+Stringtie",expression(paste("TOGA+Stringtie+",BRAKER[protein])))
toga1st_busco <- data %>% filter(method %in% c("toga","stiestar","bprot","togastie","togastiebprot")) %>%
      mutate(method=factor(method,levels=c("toga","stiestar","bprot","togastie","togastiebprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=busco_score,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","magenta","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("BUSCO score") + xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(toga1st_busco)
ggsave(toga1st_busco,file="Figure_S17A.pdf",width=4,height=2.5)
```

### Figure S17B: number of genes with NCBI BLASTP hits
```{r}
toga1st_genesblast<- data %>% filter(method %in% c("toga","stiestar","bprot","togastie","togastiebprot")) %>%
      mutate(method=factor(method,levels=c("toga","stiestar","bprot","togastie","togastiebprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=number_genes_blasthits,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","magenta","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
       ylab("Genes with BLASTP hits to NCBI") + scale_y_continuous(expand=c(0,0)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(toga1st_genesblast)
ggsave(toga1st_genesblast,file="Figure_S17B.pdf",width=4,height=3)
```

### Figure S17C: proportion genes with NCBI BLASTP hits
```{r}
toga1st_blastfrac<- data %>% filter(method %in% c("toga","stiestar","bprot","togastie","togastiebprot")) %>%
      mutate(method=factor(method,levels=c("toga","stiestar","bprot","togastie","togastiebprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=100*blastp_fraction,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","magenta","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("% of genes with NCBI BLASTP hits") +
      scale_y_continuous(expand=c(0,0)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(toga1st_blastfrac)
ggsave(toga1st_blastfrac,file="Figure_S17C.pdf",width=4,height=3)
```

### Figure S17D: RNA-seq alignment rate
```{r}
toga1st_alignrate<- data %>% filter(method %in% c("toga","stiestar","bprot","togastie","togastiebprot")) %>%
      mutate(method=factor(method,levels=c("toga","stiestar","bprot","togastie","togastiebprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=align_rate,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","magenta","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
       ylab(expression(paste(RNA[seq], " alignment rate"))) +
      scale_y_continuous(expand=c(0,0),limits=c(0,0.8)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(toga1st_alignrate)
ggsave(toga1st_alignrate,file="Figure_S17D.pdf",width=4,height=2.5)
```



## Figure S18: stringtie as base annotation, no TOGA

### Figure S18A: BUSCO scores

```{r}
methodlabels<-c(expression(Stringtie[STAR]),expression(BRAKER[RNA]),expression(BRAKER[protein]),expression(paste("Stringtie+",BRAKER[RNA])),expression(paste("Stringtie+",BRAKER[RNA],"+",BRAKER[protein])))
notoga_busco <- data %>% filter(method %in% c("stiestar","brna","bprot","stiebrna","stiebrnabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","brna","bprot","stiebrna","stiebrnabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=busco_score,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","#e69f00","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("Genes with BLASTP hits to NCBI") +
      xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(notoga_busco)
ggsave(notoga_busco,file="Figure_S18A.pdf",width=4,height=3)
```

### Figure S18B: number genes NCBI BLASTP hits
```{r}
notoga_geneblast <- data %>% filter(method %in% c("stiestar","brna","bprot","stiebrna","stiebrnabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","brna","bprot","stiebrna","stiebrnabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=number_genes_blasthits,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","#e69f00","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("Genes with BLASTP hits to NCBI") + xlab("") + scale_y_continuous(expand=c(0,0)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(notoga_geneblast)
ggsave(notoga_geneblast,file="Figure_S18B.pdf",width=4,height=3)
```


### Figure S18C: proportion genes with NCBI BLASTP hits

```{r}
notoga_blastfrac <- data %>% filter(method %in% c("stiestar","brna","bprot","stiebrna","stiebrnabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","brna","bprot","stiebrna","stiebrnabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=100*blastp_fraction,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","#e69f00","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("% of genes with NCBI BLASTP hits") +
      xlab("") + scale_y_continuous(expand=c(0,0)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(notoga_blastfrac)
ggsave(notoga_blastfrac,file="Figure_S18C.pdf",width=4,height=3)
```


### Figure S18D: RNA-seq alignment rate

```{r}
notoga_alignrate <- data %>% filter(method %in% c("stiestar","brna","bprot","stiebrna","stiebrnabprot")) %>%
      mutate(method=factor(method,levels=c("stiestar","brna","bprot","stiebrna","stiebrnabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=align_rate,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("magenta","#e69f00","yellow2","dodgerblue","blue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab(expression(paste(RNA[seq], " alignment rate"))) +
      xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,0.8)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(notoga_alignrate)
ggsave(notoga_alignrate,file="Figure_S18D.pdf",width=4,height=3)
```

## Figure S19: no RNA-seq, TOGA as base annotation
 
### Figure S19A: BUSCO scores

```{r}
methodlabels<-c("TOGA", expression(BRAKER[protein]), expression(paste("TOGA+",BRAKER[protein])))
norna_busco <- data %>% filter(method %in% c("toga","bprot","togabprot")) %>%
      mutate(method=factor(method,levels=c("toga","bprot","togabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=busco_score,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","yellow2","dodgerblue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("BUSCO score") + xlab("") + scale_y_continuous(expand=c(0,0),limits=c(0,1.05)) + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(norna_busco)
ggsave(norna_busco,file="Figure_S19A.pdf",width=4,height=3)
```

### Figure S19B: number genes NCBI BLASTP hits

```{r}
norna_geneblast <- data %>% filter(method %in% c("toga","bprot","togabprot")) %>%
      mutate(method=factor(method,levels=c("toga","bprot","togabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=number_genes_blasthits,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","yellow2","dodgerblue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("Genes with BLASTP hits to NCBI")+ 
      xlab("") + scale_y_continuous(expand=c(0,0))  + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(norna_geneblast)
ggsave(norna_geneblast,file="Figure_S19B.pdf",width=4,height=3)
```

### Figure S19C: number of genes with NCBI BLASTP hits
```{r}
norna_blastfrac <- data %>% filter(method %in% c("toga","bprot","togabprot")) %>%
      mutate(method=factor(method,levels=c("toga","bprot","togabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=100*blastp_fraction,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","yellow2","dodgerblue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab("% of genes with NCBI BLASTP hits") +
      xlab("") + scale_y_continuous(expand=c(0,0))  + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(norna_blastfrac)
ggsave(norna_blastfrac,file="Figure_S19C.pdf",width=4,height=3)
```

### Figure S19D: RNA-seq alignment rate
```{r}
norna_alignrate <- data %>% filter(method %in% c("toga","bprot","togabprot")) %>%
      mutate(method=factor(method,levels=c("toga","bprot","togabprot"))) %>%
      ggplot() + theme(text = element_text(size = 8))  +
      geom_bar(aes(method,y=align_rate,fill=method),width=0.7,stat="identity",color="black") +
      scale_fill_manual(labels=methodlabels,values=c("black","yellow2","dodgerblue")) +
      theme(axis.text.y=element_text(size=7)) + 
      ylab(expression(paste(RNA[seq], " alignment rate"))) +
      xlab("") + scale_y_continuous(expand=c(0,0))  + 
      theme(legend.text.align = 0,legend.title=element_text(size=8),legend.text=element_text(size=7),legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) + 
      guides(fill = guide_legend(byrow = TRUE)) +
      labs(fill = "Method") +
      theme(axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank())
      
print(norna_alignrate)
ggsave(norna_alignrate,file="Figure_S19D.pdf",width=4,height=3)
```


