---
  title: "Genome Annotation Project"
subtitle: "Figure S7: gini index dot plot more species"
date: "April 10, 2024"
author: "Adam Freedman"
output: 
  html_document: default
---
  
## Set up
```{r,,message=FALSE, warning=FALSE}
currentdir=getwd()
setwd(currentdir)
library(tidyverse)
``` 

## Data
```{r,,message=FALSE, warning=FALSE}
annotations <-read_csv("GenomeAnnotationMetrics_final.csv")
annotations <- annotations %>% filter(!method %in% c("BrkProt-unfilt","BrkProt-fullsupport","BrkRNA-unfilt",
                                                     "BrkRNA-fullsupport","CGPdenovo","TSEBRA","MakerNoRnaNoEvm",
                                                     "MakerRnaEvm","MakerNoRnaEvm")) %>%
                                                      filter(group!="heliconines")

genomeQC <-read_csv("GenomeQcMetrics_final.csv") 
data <- annotations<-left_join(annotations,genomeQC,by = join_by(species==Species))

data$method <- factor(data$method,levels=c("BrkProt","BrkRNA","CGPprotein","CGPrnaseq",
                                           "ScalHisat","ScalStar","StrHisat","StrStar","MakerRnaNoEvm","TOGA"))

data <- data %>% mutate(species = str_replace(species, "dmelanogaster", "D. melanogaster")) %>%
     mutate(species = str_replace(species, "dpseudoobscura", "D. pseudoobscura")) %>%  
     mutate(species = str_replace(species, "dyakuba", "D. yakuba")) %>%
     mutate(species = str_replace(species, "alyrata", "A. lyrata")) %>%
     mutate(species = str_replace(species, "athaliana", "A. thaliana")) %>%
     mutate(species = str_replace(species, "boleracea", "B. oleracea")) %>%
     mutate(species = str_replace(species, "crubella", "C. rubella")) %>%
     mutate(species = str_replace(species, "osativa", "O. sativa")) %>%
     mutate(species = str_replace(species, "sitalica", "S. italica")) %>%
     mutate(species = str_replace(species, "bdistachyon", "B. distachyon")) %>%
     mutate(species = str_replace(species, "zmays", "Z. mays")) %>%
     mutate(species = str_replace(species, "human", "H. sapiens")) %>%
     mutate(species = str_replace(species, "macaque", "M. mulatta")) %>%
     mutate(species = str_replace(species, "dog", "C. familiaris")) %>%
     mutate(species = str_replace(species, "mouse", "M. musculus")) %>%
     mutate(species = str_replace(species, "chicken", "G. gallus")) %>%
     mutate(species = str_replace(species, "quail", "C. japonica")) %>%
     mutate(species = str_replace(species, "mallard", "A. platyrhynchos"))

data <- data %>% mutate_at(vars(species),factor,levels=c("D. melanogaster","D. pseudoobscura",
                                                         "D. yakuba", "A. thaliana", "A. lyrata","B. oleracea",
                                                         "C. rubella","Z. mays","O. sativa","S. italica",
                                                         "B. distachyon", "G. gallus","A. platyrhynchos",
                                                         "C. japonica","H. sapiens","M. mulatta","C. familiaris",
                                                         "M. musculus"))

fusion_long <- pivot_longer(data, c(cds_gene_fusions,multi_cds_gene_fusions,multi_mix_gene_fusions), names_to="fusiontype", values_to="fusionfreq")
fusion_long <- fusion_long %>% mutate(fusiontype=factor(fusiontype,levels=c("cds_gene_fusions","multi_cds_gene_fusions","multi_mix_gene_fusions")))
fusion_long <- fusion_long %>% filter(species=="Z. mays" | species=="H. sapiens" | species=="D. melanogaster" |species=="A. thaliana"| species=="G. gallus")
fusion_long<- fusion_long %>% mutate(species=factor(species,levels=c("D. melanogaster","G. gallus","H. sapiens","A. thaliana","Z. mays")))
```

## Figure S8
```{r}  
fusion_labels=c("Fusion transcripts", "1 tscript-per-gene, multi-gene","Mixed")
fusion_barplot <- fusion_long %>% ggplot() + 
    geom_bar(aes(x=method,y=fusionfreq/protcod_genes,fill=fusiontype),width=0.7,stat="identity",position="stack",color="black") +
    theme(axis.text.x=element_text(angle=90, hjust=1,vjust=0.3,size=9)) +
    scale_fill_manual(labels=fusion_labels,values=c("white","dodgerblue","darkgreen")) +
    theme(legend.text.align = 0,legend.spacing.y = unit(0.15, "cm"),legend.key.size= unit(0.5, "cm")) +
    ylab("Fusion frequency") + xlab("") +
    scale_x_discrete(breaks=levels(fusion_long$method),labels=methodlabels) +
    guides(fill = guide_legend(byrow = TRUE)) +
    labs(fill = "Fusion type") +
    facet_wrap(~species) +
    theme(strip.text = element_text(face="italic"))
print(fusion_barplot)
ggsave(fusion_barplot,file="Figure_S8.pdf",width=7,height=5)
```